File: algorithmic_heights/bins/search_logic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


def search(arr, k):
    if not arr or k < arr[0] or k > arr[-1]:
        return -1

    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = (left + right) // 2
        if k == arr[mid]:
            return mid
        if k < arr[mid]:
            right = mid - 1
        else:
            left = mid + 1
    return -1


if __name__ == "__main__":
    import unittest

    class BinarySearchTestCase(unittest.TestCase):
        def test_empty_array(self):
            self.assertEqual(search([], 10), -1)

        def test_one_element_array_exclusive(self):
            self.assertEqual(search([55], 10), -1)

        def test_one_element_array_inclusive(self):
            self.assertEqual(search([10], 10), 0)

        def test_exclusive_less(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], -10), -1)

        def test_exclusive_mid(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], 79), -1)

        def test_exclusive_great(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], 999), -1)

        def test_inclusive_first(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], -5), 0)

        def test_inclusive_mid(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], 80), 4)

        def test_inclusive_last(self):
            self.assertEqual(search([-5, 6, 36, 74, 80, 238, 550], 550), 6)

    unittest.main()

File: algorithmic_heights/bins/SearchLogic.fs
module SearchLogic

let rec private search0 (arr : int[]) (k : int) (left : int) (right : int) : int =
    if (left <= right) then
        let mid = (left + right) / 2
        if k = arr.[mid] then
            mid
        elif k < arr.[mid] then
            search0 arr k left (mid - 1)
        else
            search0 arr k (mid + 1) right
    else
        -1

let search (arr : int[]) (k : int) : int =
    if k < (Array.head arr) || k > (Array.last arr) then
        -1
    else
        search0 arr k 0 (Array.length arr - 1)
